driver-core/src/test/resources/uri-options/connection-pool-options.json

```diff
@@ -28,6 +30,28 @@
       "hosts": null,
       "auth": null,
       "options": {}
+    },
+    {
+      "description": "maxPoolSize=0 does not error",
+      "uri": "mongodb://example.com/?maxPoolSize=0",
```

Comment:
I don't think our `DefaultConnectionPool` interprets `maxPoolSize=0` as "unlimited":

1. `DefaultConnectionPool.createTimeoutException` works incorrectly in this case;
2. `ConcurrentPool` is created with 0 semaphore permits in this case.

There also appears to be a bug in the constructor of the `ConcurrentPool`: it claims _"if < 0 then no limit"_ but instead creates a semaphore with negative initial permits, which means connections must be released before they can be created.

P.S. I also need to take the above into account in https://github.com/mongodb/mongo-java-driver/pull/701.

Reply:
I see.  Can you report bugs on the above behavior?  I don't see a reason not to push this PR as is, since it's not changing any driver behavior.

Reply:
Done: https://jira.mongodb.org/browse/JAVA-4216